File: backend/msvoo/src/main/java/com/ms/voo/rabbitmq/RabbitMQConfig.java
package com.ms.voo.rabbitmq;

import org.springframework.amqp.core.*;
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Bean;

@Configuration
public class RabbitMQConfig {

    private static final String DEFAULT_ROUTING_KEY = "voo";

    @Bean
    public Queue vooCancelQueue() {
        return new Queue("cancelavoo.reserva");
    }

    @Bean
    public Queue novasReservas() {
        return new Queue("criareserva.voo");
    }

    @Bean
    public DirectExchange sagaCriarReserva() {
        return new DirectExchange("criareserva");
    }

    @Bean
    public Queue vooRealizeQueue() {
        return new Queue("realizavoo.reserva");
    }

    @Bean
    public DirectExchange vooExchange() {
        return new DirectExchange("voo.exchange");
    }

    @Bean
    public Binding bindingVooCancel(DirectExchange vooExchange, Queue vooCancelQueue) {
        return BindingBuilder.bind(vooCancelQueue)
                .to(vooExchange)
                .with("cancelavoo.reserva");
    }

    @Bean
    public Binding bindingVooRealize(DirectExchange vooExchange, Queue vooRealizeQueue) {
        return BindingBuilder.bind(vooRealizeQueue)
                .to(vooExchange)
                .with("realizavoo.reserva");
    }

    @Bean
    public Binding bindingCriarReserva(DirectExchange sagaCriarReserva, Queue novasReservas) {
        return BindingBuilder.bind(novasReservas)
                .to(sagaCriarReserva)
                .with(DEFAULT_ROUTING_KEY);
    }

    @Bean
    public SimpleRabbitListenerContainerFactory rabbitListenerContainerFactory(ConnectionFactory connectionFactory) {
        SimpleRabbitListenerContainerFactory factory = new SimpleRabbitListenerContainerFactory();

        factory.setConnectionFactory(connectionFactory);
        factory.setDefaultRequeueRejected(false);

        return factory;
    }
}

File: backend/voo-service/src/main/kotlin/br/ufpr/dac/voo_service/RabbitMQConfig.kt
package br.ufpr.dac.voo_service

import org.springframework.amqp.core.Binding
import org.springframework.amqp.core.BindingBuilder
import org.springframework.amqp.core.DirectExchange
import org.springframework.amqp.core.Queue
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class RabbitMQConfig {
    private val DEFAULT_ROUTING_KEY = "voo"

    @Bean
    fun novasReservas(): Queue = Queue("emiratads.criareserva.voo")

    @Bean
    fun sagaCriarReserva(): DirectExchange = DirectExchange("emiratads.criareserva")

    @Bean
    fun cancelarReserva(): Queue = Queue("emiratads.cancelareserva.voo")

    @Bean
    fun sagaCancelarReserva(): DirectExchange = DirectExchange("emiratads.cancelareserva")

    @Bean
    fun cancelarVoo(): Queue = Queue("emiratads.cancelavoo.voo")

    @Bean
    fun sagaCancelarVoo(): DirectExchange = DirectExchange("emiratads.cancelavoo")

    @Bean
    fun realizarVoo(): Queue = Queue("emiratads.realizavoo.voo")

    @Bean
    fun sagaRealizarVoo(): DirectExchange = DirectExchange("emiratads.realizavoo")

    @Bean
    fun bindingCriarReserva(
        sagaCriarReserva: DirectExchange,
        novasReservas: Queue
    ): Binding {
        return BindingBuilder.bind(novasReservas)
            .to(sagaCriarReserva)
            .with(DEFAULT_ROUTING_KEY)
    }

    @Bean
    fun bindingCancelarReserva(
        sagaCancelarReserva: DirectExchange,
        cancelarReserva: Queue
    ): Binding {
        return BindingBuilder.bind(cancelarReserva)
            .to(sagaCancelarReserva)
            .with(DEFAULT_ROUTING_KEY)
    }

    @Bean
    fun bindingCancelarVoo(
        sagaCancelarVoo: DirectExchange,
        cancelarVoo: Queue
    ): Binding {
        return BindingBuilder.bind(cancelarVoo)
            .to(sagaCancelarVoo)
            .with(DEFAULT_ROUTING_KEY)
    }

    @Bean
    fun bindingRealizarVoo(
        sagaRealizarVoo: DirectExchange,
        realizarVoo: Queue
    ): Binding {
        return BindingBuilder.bind(realizarVoo)
            .to(sagaRealizarVoo)
            .with(DEFAULT_ROUTING_KEY)
    }

    @Bean
    fun rabbitListenerContainerFactory(connectionFactory: ConnectionFactory): SimpleRabbitListenerContainerFactory {
        val factory = SimpleRabbitListenerContainerFactory();
        factory.setConnectionFactory(connectionFactory)
        factory.setDefaultRequeueRejected(false)
        return factory
    }
}